lib/active_merchant/billing/gateways/redsys.rb

```diff
@@ -321,6 +323,26 @@ def add_payment(data, card)
         end
       end
 
+      def add_external_mpi_fields(data, options)
+        return unless options[:three_d_secure]
+
+        if options[:three_d_secure][:version]
+          data[:threeDSServerTransID] = options[:three_d_secure][:xid] if options[:three_d_secure][:xid]
+          data[:dsTransID] = options[:three_d_secure][:ds_transaction_id] if options[:three_d_secure][:ds_transaction_id]
+          data[:authenticacionValue] = options[:three_d_secure][:cavv] if options[:three_d_secure][:cavv]
+          data[:protocolVersion] = options[:three_d_secure][:version] if options[:three_d_secure][:version]
+
+          # TODO: keep here?
```

Comment:
Were you considering pulling these out to another method? I think they're alright here.

Reply:
Leila suggested these could actually live as optional via our gateway integration in`core`. I welcome your thoughts around that too

Reply:
I see. Yes, I think these non-standard but 3ds-related fields should be general options rather than assume dto exist in the `three_d_secure` hash.